apis/Google.Devtools.AspNet/Google.Devtools.AspNet.Tests/CloudTrace/GrpcTraceConsumerTest.cs

```diff
@@ -30,39 +31,78 @@ public class GrpcTraceConsumerTest
         [Fact]
         public void Receive()
         {
+            ManualResetEvent reset = new ManualResetEvent(false);
             Traces traces = GetTraces();
 
             Mock<TraceServiceClient> mockClient = new Mock<TraceServiceClient>();
-            mockClient.Setup(c => c.PatchTracesAsync(ProjectId, traces, null));
+            mockClient.Setup(c => c.PatchTracesAsync(ProjectId, traces, null)).Callback(() => reset.Set());
             Task<TraceServiceClient> taskClient = Task.FromResult(mockClient.Object);
             GrpcTraceConsumer consumer = new GrpcTraceConsumer(taskClient);
 
             consumer.Receive(traces);
+            Utils.WaitForSet(reset, "PatchTracesAsync wasn't called");
             mockClient.VerifyAll();
         }
 
         [Fact]
         public void Receive_EmptyTracesIgnored()
         {
+            ManualResetEvent reset = new ManualResetEvent(false);
             Mock<TraceServiceClient> mockClient = new Mock<TraceServiceClient>();
+            mockClient.Setup(c => c.PatchTracesAsync(It.IsAny<string>(), It.IsAny<Traces>(), null))
+                .Callback(() => reset.Set());
             Task<TraceServiceClient> taskClient = Task.FromResult(mockClient.Object);
             GrpcTraceConsumer consumer = new GrpcTraceConsumer(taskClient);
 
             consumer.Receive(new Traces());
+            Utils.EnsureNoSet(reset, "PatchTracesAsync was called");
             mockClient.Verify(c => c.PatchTracesAsync(It.IsAny<string>(), It.IsAny<Traces>(), null), Times.Never());
         }
 
         [Fact]
         public void Receive_ClientNotReady()
         {
+            ManualResetEvent reset = new ManualResetEvent(false);
             Traces traces = GetTraces();
 
             Mock<TraceServiceClient> mockClient = new Mock<TraceServiceClient>();
+            mockClient.Setup(c => c.PatchTracesAsync(It.IsAny<string>(), It.IsAny<Traces>(), null))
+                .Callback(() => reset.Set());
             Task<TraceServiceClient> taskClient = new Task<TraceServiceClient>(() => mockClient.Object);
             GrpcTraceConsumer consumer = new GrpcTraceConsumer(taskClient);
 
             consumer.Receive(traces);
+            Utils.EnsureNoSet(reset, "PatchTracesAsync was called");
             mockClient.Verify(c => c.PatchTracesAsync(It.IsAny<string>(), It.IsAny<Traces>(), null), Times.Never());
         }
+
+        [Fact]
+        public void Receive_TooManyPendingTasks()
+        {
+            Traces traces = GetTraces();
+            Mock<TraceServiceClient> mockClient = new Mock<TraceServiceClient>();
+            mockClient.Setup(c => c.PatchTracesAsync(It.IsAny<string>(), It.IsAny<Traces>(), null));
+            Task<TraceServiceClient> taskClient = new Task<TraceServiceClient>(() => mockClient.Object);
```

Comment:
Just use a TaskCompletionSource, and call SetResult later instead of RunSynchronously. That's the simplest way to have a task you control.